enhancements/cloud-integration/openstack/cloud-provider-config-upgrade.md

```diff
@@ -0,0 +1,441 @@
+to the user-managed config and writes the result to
+`openshift-config-managed/kube-cloud-config`. The transformation
+function is selected according to PlatformType in the cluster Infrastructure
+object. Components requiring legacy cloud provider configuration, including
+kubelet and KCM, read the transformed configmap.
+
+CCO will **not** be modified. The new functionality will instead be implemented
+as a change to the [Cluster Cloud Controller Manager Operator (CCCMO)][CCCMO],
+functioning in much the same way as CCO's management of the legacy
+configuration.
+
+[CCO]: https://docs.openshift.com/container-platform/4.9/operators/operator-reference.html#cluster-config-operator_platform-operators-ref
+[CCCMO]: https://docs.openshift.com/container-platform/4.9/operators/operator-reference.html#cluster-cloud-controller-manager-operator_platform-operators-ref
+
+### Separate ConfigMaps for legacy and external cloud providers
+
+The cloud configuration is read by multiple components which will not all be
+upgraded simultaneously. If we simply updated the
+`openshift-config-managed/kube-cloud-config` to be compatible with external CCM
+we would risk breaking KCM and kubelets which have not yet been upgraded.
+Therefore we are constrained to write two versions of the managed config map.
+
+CCO will be continue to create `openshift-config-managed/kube-cloud-config`,
+which will remain compatible with the legacy cloud provider. CCCMO will create a
+new `openshift-config-managed/cloud-controller-manager-config` config map, which
+will be compatible with the external CCM. The external CCM will then read its
+configuration from `openshift-config-managed/cloud-controller-manager-config`.
+
+In a future release, we can deprecate and remove CCO's management of
+`openshift-config-managed/kube-cloud-config` and eventually the config map itself.
+
+### Preventing upgrade when config cannot be upgraded
+
+If CCCMO is not able to generate `cloud-controller-manager-config` from
+`openshift-config/cloud-provider-config`, it will mark itself as non-upgradable
+with a user-actionable error message explaining how
+`openshift-config/cloud-provider-config` is incompatible.
+
+As we intend to switch to external CCM in 4.11, this change will be implemented
+in OCP 4.10.
+
+### Compatibility constraints on user-managed config
+
+This has implications for the user-managed configuration:
+
+In 4.10, the user-managed configuration **must** be compatible with the legacy
+cloud provider. The user **must not** specify config options which cannot be
+transformed to config for the external cloud provider. If either of these are
+not satisfied we will mark CCCMO as not upgradeable.
+
+In 4.11, the base `cloud-provider-config` **may** include config options which
+are not compatible with the legacy cloud provider. If these options cannot be
+transformed for the legacy cloud provider we will not update
+`kube-cloud-config`, but will not delete the existing one.
```

Comment:
This suggests we'll keep syncing the legacy cloud config even after we've switched to the CCM, what is the point?
We explicitly say "Downgrade from CCM is not supported" below.

Reply:
Good point. Done